src/main/java/com/commercetools/sync/services/impl/BaseService.java

```diff
@@ -0,0 +1,50 @@
+package com.commercetools.sync.services.impl;
+
+import com.commercetools.sync.commons.BaseSyncOptions;
+import com.commercetools.sync.commons.utils.TriFunction;
+import io.sphere.sdk.commands.DraftBasedCreateCommand;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+class BaseService<U, V> {
```

Comment:
I'd limit the lower generic types bounds:
`BaseService<U extends io.sphere.sdk.models.Resource<U>, V>`

So in the future we could access `U` values  from base class for some basic features (`id`, `version`, `toReference`).

Reply:
Java doc would be enough for me.

Reply:
here is question is not about documentation, but about design: it is usually recommended to keep the type as narrow as possible, so in future extending the functionality is less painful, e.g. we limit that the `BaseService` is intended to works only with CTP Resource Entities (Products, Categories, Inventories etc), but not with other types (Strings, Socket connections whatever)

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/199/commits/a90e5788e870b3362f920fc556b1185d977f2029